### DISCUSSION TITLE
Leader Election With S3 Conditional Writes - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#319] [Announcements]
# Leader Election With S3 Conditional Writes - Gunnar Morling

Update Aug 30: This article is discussed on Hacker News and lobste.rs.

In distributed systems, for instance when scaling out some workload to multiple compute nodes,
it is a common requirement to select a leader for performing a given task:
only one of the nodes should process the records from a Kafka topic partition, write to a file system, call a remote API, etc.
Otherwise, multiple workers may end up doing the same task twice, overwriting each other’s data, and worse.

https://www.morling.dev/blog/leader-election-with-s3-conditional-writes/

**Comment by spoiledwit** [comment]
Great post

**Reply by gunnarmorling** [reply]
Thank you, appreciate it!

**Comment by itsKarad** [comment]
Great post, acted as a good refresher to my distributed systems course!

**Comment by clayg** [comment]
Does s3 serialize concurrent writes to guarantee only one creator?  Has anyone tested doing an “x-if-none-match: *” for an object that doesn’t exist while another  unfinished PUT is in flight?  Is that kind of serialization required for this locking algorithm?

**Comment by gunnarmorling** [comment]
Good question; this sort of scenario is what conditional writes help to resolve. I just gave it a try by concurrently uploading two files for the same key, one with a size of 1MB, one with 100MB.

In session 1:

```
mkfile 100m large_file.bin
aws s3api put-object --bucket my-bucket --key my-file.bin \
    --body large_file.bin --if-none-match "*"

# The file upload takes a while...
```


In session 2:

```
mkfile 1m small_file.bin
aws s3api put-object --bucket my-bucket --key my-file.bin \
    --body small_file.bin --if-none-match "*"

{
    "ETag": "\"b6d81b360a5672d80c27430f39153e2c\"",
    "ServerSideEncryption": "AES256"
}
```

Finally in session 1, when the file has been fully transmitted:

```
An error occurred (PreconditionFailed) when calling the PutObject operation:
At least one of the pre-conditions you specified did not hold
```

I.e. the last writer to finish is rejected.

**Reply by clayg** [reply]
Fantastic. So it’s like a two-phase commit. I assume attempting to re-upload the 100MB object with if-none-match is rejected immediately.

**Reply by gunnarmorling** [reply]
That's what I would have thought too, but indeed it transmits the entire file before failing.

**Comment by tinin007k** [comment]
Great post. The markdown file 0001-manifest.md seems to have moved to https://github.com/slatedb/slatedb/blob/main/rfcs/0001-manifest.md.

**Reply by gunnarmorling** [reply]
Thx, good catch! Updated it.
